apiserver/facades/client/spaces/remove.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package spaces
+
+import (
+	"github.com/juju/errors"
+	"gopkg.in/juju/names.v3"
+	"gopkg.in/mgo.v2/txn"
+
+	"github.com/juju/juju/apiserver/common"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/controller"
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/core/permission"
+)
+
+// RemoveSpace describes a space that can be removed.
+type RemoveSpace interface {
+	Refresh() error
+	RemoveSpaceOps() []txn.Op
+}
+
+type spaceRemoveModelOp struct {
+	space   RemoveSpace
+	subnets []Subnet
+}
+
+type Subnet interface {
+	MoveSubnetOps(spaceID string) []txn.Op
+}
+
+func (o *spaceRemoveModelOp) Done(err error) error {
+	return err
+}
+
+func NewRemoveSpaceModelOp(space RemoveSpace, subnets []Subnet) *spaceRemoveModelOp {
+	return &spaceRemoveModelOp{
+		space:   space,
+		subnets: subnets,
+	}
+}
+
+func (sp *spaceRemoveModelOp) Build(attempt int) ([]txn.Op, error) {
+	var totalOps []txn.Op
+
+	if attempt > 0 {
+		if err := sp.space.Refresh(); err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	for _, subnet := range sp.subnets {
+		totalOps = append(totalOps, subnet.MoveSubnetOps(network.AlphaSpaceId)...)
+	}
+
+	removeOps := sp.space.RemoveSpaceOps()
+	totalOps = append(totalOps, removeOps...)
+	return totalOps, nil
+}
+
+// RemoveSpace removes a space.
+// Returns SpaceResults if entities/settings are found which makes the deletion not possible.
+func (api *API) RemoveSpace(entities params.Entities) (params.RemoveSpaceResults, error) {
+	isAdmin, err := api.auth.HasPermission(permission.AdminAccess, api.backing.ModelTag())
+	if err != nil && !errors.IsNotFound(err) {
+		return params.RemoveSpaceResults{}, errors.Trace(err)
+	}
+	if !isAdmin {
+		return params.RemoveSpaceResults{}, common.ServerError(common.ErrPerm)
+	}
+	if err := api.check.ChangeAllowed(); err != nil {
+		return params.RemoveSpaceResults{}, errors.Trace(err)
+	}
+	if err = api.checkSupportsProviderSpaces(); err != nil {
+		return params.RemoveSpaceResults{}, common.ServerError(errors.Trace(err))
+	}
+
+	results := params.RemoveSpaceResults{
+		Results: make([]params.RemoveSpaceResult, len(entities.Entities)),
+	}
+	for i, entity := range entities.Entities {
```

Comment:
If you extract the body of this loop into a method that returns an error, you can trim down the repetition of error assignment + continue.